### ISSUE TITLE
FindPyArrow.cmake not accessing environment variables correctly

### ISSUE THREAD

**Issue by sjperkins** [ratt-ru/arcae#78]
As discovered by @orliac, the following doesn't seem to access environment variables correctly:

https://github.com/ratt-ru/arcae/blob/bebaf525c29c76de6dd7e1951b65359943f41e9f/cmake_modules/FindPyArrow.cmake#L37-L39

and should probably be something like

```cpp
if(NOT DEFINED PYARROW_LIBDIRS AND DEFINED ENV{PYARROW_LIBDIRS})
    set(PYARROW_LIBDIRS $ENV{PYARROW_LIBDIRS})
endif()
```

**Comment by orliac** [comment]
Also, in my environment I had to change
```
if(NOT DEFINED PYARROW_LIBDIRS AND Python_Interpreter_FOUND)
    execute_process(COMMAND ${Python_EXECUTABLE} -c "import pyarrow; print(' '.join(pyarrow.get_library_dirs()))"
                    OUTPUT_VARIABLE PYARROW_LIBDIRS
                    OUTPUT_STRIP_TRAILING_WHITESPACE)

endif()
```
for
```
if(NOT DEFINED PYARROW_LIBDIRS AND Python_Interpreter_FOUND)
    execute_process(COMMAND ${Python_EXECUTABLE} -c "import pyarrow; print(pyarrow.get_library_dirs()[0])"
                    OUTPUT_VARIABLE PYARROW_LIBDIRS
                    OUTPUT_STRIP_TRAILING_WHITESPACE)
endif()
```
In the first case `PYARROW_LIBDIRS = /tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/arrow-13.0.0-moz5ce63c57ndqus43akxvfm3z7jh3hr/lib64` and compilation fails with:

```
(VENV2) [rb-dev0] [orliac@i63 arcae]$ pip -v install -e .
Using pip 23.1.2 from /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/lib/python3.11/site-packages/pip (python 3.11)
Obtaining file:///home/orliac/RADIOBLOCKS/testing_arcae/arcae
  Running command pip subprocess to install build dependencies
  Collecting cython>=3.0.2
    Using cached Cython-3.0.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.6 MB)
  Collecting scikit-build-core
    Using cached scikit_build_core-0.6.1-py3-none-any.whl (134 kB)
  Collecting pyarrow==13.0.0
    Using cached pyarrow-13.0.0-cp311-cp311-manylinux_2_28_x86_64.whl (40.0 MB)
  Collecting numpy>=1.16.6 (from pyarrow==13.0.0)
    Using cached numpy-1.26.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)
  Collecting packaging>=20.9 (from scikit-build-core)
    Using cached packaging-23.2-py3-none-any.whl (53 kB)
  Installing collected packages: packaging, numpy, cython, scikit-build-core, pyarrow
  ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
  numba 0.57.0 requires numpy<1.25,>=1.21, but you have numpy 1.26.2 which is incompatible.
  Successfully installed cython-3.0.5 numpy-1.26.2 packaging-23.2 pyarrow-13.0.0 scikit-build-core-0.6.1

  [notice] A new release of pip is available: 23.1.2 -> 23.3.1
  [notice] To update, run: pip install --upgrade pip
  Installing build dependencies ... done
  Running command Checking if build backend supports build_editable
  Checking if build backend supports build_editable ... done
  Running command Getting requirements to build editable
  Getting requirements to build editable ... done
  Running command pip subprocess to install backend dependencies
  Collecting pyproject_metadata
    Using cached pyproject_metadata-0.7.1-py3-none-any.whl (7.4 kB)
  Collecting pathspec
    Using cached pathspec-0.11.2-py3-none-any.whl (29 kB)
  Collecting packaging>=19.0 (from pyproject_metadata)
    Using cached packaging-23.2-py3-none-any.whl (53 kB)
  Installing collected packages: pathspec, packaging, pyproject_metadata
  Successfully installed packaging-23.2 pathspec-0.11.2 pyproject_metadata-0.7.1

  [notice] A new release of pip is available: 23.1.2 -> 23.3.1
  [notice] To update, run: pip install --upgrade pip
  Installing backend dependencies ... done
  Running command Preparing editable metadata (pyproject.toml)
  *** scikit-build-core 0.6.1 using CMake 3.27.7 (metadata_editable)
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: appdirs in ./VENV2/lib/python3.11/site-packages (from arcae==0.2.1) (1.4.4)
Requirement already satisfied: click in /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/lib/python3.11/site-packages (from arcae==0.2.1) (8.1.7)
Requirement already satisfied: rich in ./VENV2/lib/python3.11/site-packages (from arcae==0.2.1) (13.6.0)
Requirement already satisfied: pyarrow==13.0.0 in /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/lib/python3.11/site-packages (from arcae==0.2.1) (13.0.0)
Requirement already satisfied: numpy>=1.16.6 in /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/lib/python3.11/site-packages (from pyarrow==13.0.0->arcae==0.2.1) (1.24.3)
Requirement already satisfied: markdown-it-py>=2.2.0 in ./VENV2/lib/python3.11/site-packages (from rich->arcae==0.2.1) (3.0.0)
Requirement already satisfied: pygments<3.0.0,>=2.13.0 in ./VENV2/lib/python3.11/site-packages (from rich->arcae==0.2.1) (2.16.1)
Requirement already satisfied: mdurl~=0.1 in ./VENV2/lib/python3.11/site-packages (from markdown-it-py>=2.2.0->rich->arcae==0.2.1) (0.1.2)
Building wheels for collected packages: arcae
  Running command Building editable for arcae (pyproject.toml)
  *** scikit-build-core 0.6.1 using CMake 3.27.7 (editable)
  *** Configuring CMake...
  loading initial cache file /tmp/tmpccb85_tn/build/CMakeInit.txt
  -- The CXX compiler identification is GNU 11.3.0
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /ssoft/spack/syrah/v1/opt/spack/linux-rhel8-x86_64_v2/gcc-8.5.0/gcc-11.3.0-yrewh277xaqlocj7zcuxnv2m4fq4p3tw/bin/g++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Found Python3: /home/orliac/RADIOBLOCKS/testing_arcae/arcae/VENV2/bin/python (found version "3.11.6") found components: Interpreter
  -- Found Python: /home/orliac/RADIOBLOCKS/testing_arcae/arcae/VENV2/bin/python (found version "3.11.6") found components: Interpreter
  Python /home/orliac/RADIOBLOCKS/testing_arcae/arcae/VENV2/bin/python
  @@@ PYARROW_LIBDIRS = /tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/arrow-13.0.0-moz5ce63c57ndqus43akxvfm3z7jh3hr/lib64
  NUMPY_INCLUDE:/tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/numpy/core/include
  PYARROW_INCLUDE:/tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow/include
  PYARROW_LIBDIRS:/tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/arrow-13.0.0-moz5ce63c57ndqus43akxvfm3z7jh3hr/lib64
  PYARROW_LIBS:arrow_python arrow
  -- Found PkgConfig: /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/bin/pkg-config (found version "1.9.5")
  -- Checking for module 'casacore'
  --   Found casacore, version 3.5.0
  -- Found Python: /home/orliac/RADIOBLOCKS/testing_arcae/arcae/VENV2/bin/python (found version "3.11.6") found components: Interpreter Development.Module
  Linking arrow_tables against arrow_python
  Linking arrow_tables against arrow
  -- Configuring done (2.8s)
  -- Generating done (0.0s)
  -- Build files have been written to: /tmp/tmpccb85_tn/build
  *** Building project with Ninja...
  [1/19] Building CXX object cpp/tests/CMakeFiles/test_utils.dir/test_utils.cc.o
  [2/19] Linking CXX static library cpp/tests/libtest_utils.a
  [3/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/utility.cc.o
  [4/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/service_locator.cc.o
  [5/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/configuration.cc.o
  [6/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/complex_type.cc.o
  [7/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/casa_visitors.cc.o
  [8/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/descriptor.cc.o
  [9/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/table_factory.cc.o
  [10/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/safe_table_proxy.cc.o
  [11/19] Building CXX object cpp/tests/CMakeFiles/basic_test.dir/basic_test.cc.o
  [12/19] Building CXX object cpp/tests/CMakeFiles/parallel_write_test.dir/parallel_write_test.cc.o
  [13/19] Building CXX object cpp/CMakeFiles/arcae.dir/arcae/column_convert_visitor.cc.o
  [14/19] Linking CXX shared library cpp/libarcae.so
  FAILED: cpp/libarcae.so
  : && /ssoft/spack/syrah/v1/opt/spack/linux-rhel8-x86_64_v2/gcc-8.5.0/gcc-11.3.0-yrewh277xaqlocj7zcuxnv2m4fq4p3tw/bin/g++ -fPIC -O3 -DNDEBUG   -shared -Wl,-soname,libarcae.so -o cpp/libarcae.so cpp/CMakeFiles/arcae.dir/arcae/casa_visitors.cc.o cpp/CMakeFiles/arcae.dir/arcae/column_convert_visitor.cc.o cpp/CMakeFiles/arcae.dir/arcae/complex_type.cc.o cpp/CMakeFiles/arcae.dir/arcae/configuration.cc.o cpp/CMakeFiles/arcae.dir/arcae/descriptor.cc.o cpp/CMakeFiles/arcae.dir/arcae/service_locator.cc.o cpp/CMakeFiles/arcae.dir/arcae/safe_table_proxy.cc.o cpp/CMakeFiles/arcae.dir/arcae/table_factory.cc.o cpp/CMakeFiles/arcae.dir/arcae/utility.cc.o -L"/tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/arrow-13.0.0-moz5ce63c57ndqus43akxvfm3z7jh3hr/lib64" -Wl,-rpath,"/tmp/pip-build-env-podhq9xj/overlay/lib/python3.11/site-packages/pyarrow /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/arrow-13.0.0-moz5ce63c57ndqus43akxvfm3z7jh3hr/lib64:/work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib::::::::"  -larrow  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_casa.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_tables.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_scimath_f.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_scimath.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_measures.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_meas.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_fits.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_ms.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_msfits.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_derivedmscal.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_lattices.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_mirlib.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_coordinates.so  /work/ska/soft/spack/radioblocks/rb-dev0/opt/spack/linux-rhel8-cascadelake/gcc-11.3.0/casacore-3.5.0-yuo3i7z2ulmamk3g4hm35725k2hygbt6/lib/libcasa_images.so && :
  /work/ska/soft/spack/blackhole/v2/opt/view__rb-dev0/bin/ld: cannot find -larrow: No such file or directory
  collect2: error: ld returned 1 exit status
  ninja: build stopped: subcommand failed.

```

**Comment by sjperkins** [comment]
> Also, in my environment I had to change

I ran into this while testing locally. In my case there were artifacts remaining from a pyarrow 13.0.0  install interacting with the pyarrow 14.0.1 install. I just deleted the `lib/python3.10/site-packages/pyarrow` directory, reinstalled pyarrow 14.0.1 and the above error message no longer appeared.
